### PR TITLE
Feature Chroma support

### PR DESCRIPTION
**Pull request by kohya-ss**
Thank you lodestones for the high quality model.

`--model_type` option is added to `flux_train_network.py`. For Chroma, `--model_type chroma` is required.

According to lodestones, `--apply_t5_attn_mask --timestep_sampling sigmoid --model_prediction_type raw --guidance_scale 0.0` options are also required for Chroma. 
 
The prompts for sample image generation requires `--l 4.0 --g 0.0`  options. `--l 4.0` means CFG scale=4, `--g 0.0` means embedded guidance scale=0. And the negative prompt like `--n low quality, ugly...` is also required unlike FLUX.1.

### CODE REVIEW

**Comment by kohya-ss** [comment]
It seems that the trained LoRA can be used in ComfyUI.

**Comment by kohya-ss** [comment]
Inference with `flux_minimal_inference.py`:

```
python flux_minimal_inference.py --ckpt path/to/chroma.safetensors  --t5xxl path/to/t5xxl.safetensors 
    --ae path/to/flux_ae.safetensors --out path/to/output_dir --flux_dtype bfloat16 --offload --steps 26 
    --prompt "positive prompt" --negative "negative prompt" 
    --seed 666 --cfg_scale 4 --guidance 0 --width 1024 --height 1024 --model_type chroma 
    --apply_t5_attn_mask
```

`--dtype fp8` casts the model to fp8 and reduces memory usage.

LoRA can be specified with `--lora_weights path/to/lora.safetensors` option.

**Comment by maybleMyers** [comment]
This command worked on a turing rtx 8000. It did start producing some horizontal lines around epoch 11 but the likeness was good, I know some other trainers were having problems with lines as well.  accelerate launch --gpu_ids 0 --mixed_precision no --num_cpu_threads_per_process 10 flux_train_network.py --pretrained_model_name_or_path models/chroma-unlocked-v45-detail-calibrated.safetensors --t5xxl models/t5xxl_flan_latest.safetensors --ae models/ae.safetensors --dataset_config models/mayble/mayble.toml --output_dir models/mayble/model --output_name Mayble --save_model_as safetensors --save_precision float --max_train_epochs 100 --save_every_n_epochs 1 --network_module networks.lora_flux --network_dim 128 --network_alpha 64 --network_train_unet_only --optimizer_type AdamW --learning_rate 5e-4 --cache_latents_to_disk --cache_text_encoder_outputs --cache_text_encoder_outputs_to_disk --persistent_data_loader_workers --max_data_loader_n_workers 2 --seed 42 --gradient_checkpointing --sdpa --model_type chroma --apply_t5_attn_mask --timestep_sampling sigmoid --model_prediction_type raw --guidance_scale 0.0 --max_grad_norm 1.0 --blocks_to_swap 4

**Comment by maybleMyers** [comment]
Adding this to the launch command got rid of the lines and made quality much better:
--network_args "lr_if_contains=double_blocks$$0$$:0" "lr_if_contains=double_blocks$$1$$:0" "lr_if_contains=double_blocks$$2$$:0.00456" "lr_if_contains=double_blocks$$3$$:0.006" "lr_if_contains=double_blocks$$4$$:0.00722" "lr_if_contains=double_blocks$$5$$:0.00822" "lr_if_contains=double_blocks$$6$$:0.009" "lr_if_contains=double_blocks$$7$$:0.00956" "lr_if_contains=double_blocks$$8$$:0.00989" "lr_if_contains=double_blocks$$9$$:0.01" "lr_if_contains=double_blocks$$10$$:0.00989" "lr_if_contains=double_blocks$$11$$:0.00956" "lr_if_contains=double_blocks$$12$$:0.009" "lr_if_contains=double_blocks$$13$$:0.00822" "lr_if_contains=double_blocks$$14$$:0.00722" "lr_if_contains=double_blocks$$15$$:0.006" "lr_if_contains=double_blocks$$16$$:0.00456" "lr_if_contains=double_blocks$$17$$:0" "lr_if_contains=double_blocks$$18$$:0"

**Comment by kohya-ss** [comment]
Unfortunately sd-scripts doesn't support `lr_if_contains` network args for now. However, LyCORIS etc. may support it.

**Comment by kohya-ss** [comment]
This PR will be merged after #2138, because it includes little updates for common utils.

Edit: #2138 has been merged.

**Comment by kohya-ss** [comment]
The regex based dim (rank) and learning rate configuration has been added. See https://github.com/kohya-ss/sd-scripts/blob/feature-chroma-support/docs/flux_train_network.md#65-regular-expression-based-ranklr-configuration--%E6%AD%A3%E8%A6%8F%E8%A1%A8%E7%8F%BE%E3%81%AB%E3%82%88%E3%82%8B%E3%83%A9%E3%83%B3%E3%82%AF%E5%AD%A6%E7%BF%92%E7%8E%87%E3%81%AE%E6%8C%87%E5%AE%9A for details.

File: library/chroma_models.py [hunk 1]
```diff
@@ -695,6 +695,7 @@ def forward(
                 input_vec = self.get_input_vec(timesteps, guidance, img.shape[0])
 
                 # kohya-ss: I'm not sure why requires_grad is set to True here
+                # original code: https://github.com/lodestone-rock/flow/blob/c76f63058980d0488826936025889e256a2e0458/src/models/chroma/model.py#L217
```
Comment (rockerBOO) [inline library/chroma_models.py:698]: Maybe the other part of the comment might help here, if it is resolved or if we find a way to fix it

https://github.com/lodestone-rock/flow/blob/c76f63058980d0488826936025889e256a2e0458/src/models/chroma/model.py#L195C1-L200C45
